### ISSUE TITLE
[BUG] Classes services being constructed even if not called

### ISSUE THREAD

**Issue by noniagriconomie** [symfony/symfony#35882]
**Symfony version(s) affected**: 3.4.37 env dev PHP 7.2.27

**Description**  
I do have a symfony command that initialized (constructs) classes as services even if those are not at all called directly inside this cli neither via classes dependencies of this cli.
It occurs as well on a simple controller call.

**How to reproduce**  
For me it is either a bad configuration of my project/services or a bug. but I can not find what and why this happened

**Possible Solution**  
On slack support channel where I asked first, I was given this link https://symfony.com/doc/current/service_container/lazy_services.html but I do not think it is a valid solution as on my other projects, I do not need this kind of feature and it does not behave like this.

**Additional context**  
On other projects v4.4 and v5.0 (env dev and up to PHP 7.4), calling this command works as intented: it does not initiliazed others services non related.

---

Hi,

Above is the issue description I can give so far.
I can of course provide some code or symfony and container configuration if needed.

Thank you,

**Comment by dmaicher** [comment]
Based on that description alone and without any examples or a reproducer it will be impossible to help you with your problem I'm afraid :confused:

**Comment by noniagriconomie** [comment]
hi @dmaicher yes here are some code

**the simple command**

```php
<?php

namespace AdminBundle\Command\Admin;

use Component\Utils\StringUtils;
use CoreBundle\Component\Mailer\UserMailer;
use CoreBundle\Entity\Core\User;
use Doctrine\ORM\EntityManagerInterface;
use Symfony\Component\Console\Command\Command;
use Symfony\Component\Console\Input\InputArgument;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Output\OutputInterface;
use Symfony\Component\Console\Question\ConfirmationQuestion;
use Symfony\Component\Security\Core\Encoder\UserPasswordEncoderInterface;

class AdminUpdatePasswordCommand extends Command
{
    protected static $defaultName = 'app:admin:update-password';
    protected $em;
    protected $userEncoder;
    protected $mailer;

    public function __construct(EntityManagerInterface $em, UserPasswordEncoderInterface $userEncoder, UserMailer $mailer)
    {
        $this->em = $em;
        $this->userEncoder = $userEncoder;
        $this->mailer = $mailer;

        parent::__construct();
    }

    protected function configure()
    {
        $this->setName(self::$defaultName)
            ->setDescription('Update an admin password.')
            ->addArgument(
                'email',
                InputArgument::REQUIRED,
                'Admin email'
            );
    }

    protected function execute(InputInterface $input, OutputInterface $output)
    {
        $email = $input->getArgument('email');
        $user = $this->em->getRepository(User::class)->findOneByEmail($email);
        $userNewPass = StringUtils::generatePassword(10);
        $user->setPassword($this->userEncoder->encodePassword($user, $userNewPass));
        $this->em->flush($user);
        $this->mailer->sendAdminUpdatePasswordEmployees($user, $userNewPass);

        return 0;
    }
}


```

```bash
symfony console debug:container "DevBundle\Command\DevCommand"

Information for Service "DevBundle\Command\DevCommand"
======================================================

 ---------------- ------------------------------ 
  Option           Value                         
 ---------------- ------------------------------ 
  Service ID       DevBundle\Command\DevCommand  
  Class            DevBundle\Command\DevCommand  
  Tags             console.command               
  Calls            setName                       
  Public           no                            
  Synthetic        no                            
  Lazy             no                            
  Shared           yes                           
  Abstract         no                            
  Autowired        yes                           
  Autoconfigured   yes                           
 ---------------- ------------------------------ 
```

And I have defined some classes as services

for example:

```bash
symfony console debug:container "CoreBundle\Component\Proxy\XYZ\XYZApiProxy"

Information for Service "CoreBundle\Component\Proxy\XYZ\XYZApiProxy"
========================================================================

 ---------------- ------------------------------------------------ 
  Option           Value                                           
 ---------------- ------------------------------------------------ 
  Service ID       CoreBundle\Component\Proxy\XYZ\XYZApiProxy  
  Class            CoreBundle\Component\Proxy\XYZ\XYZApiProxy  
  Tags             -                                               
  Calls            setLogger                                       
  Public           no                                              
  Synthetic        no                                              
  Lazy             no                                              
  Shared           yes                                             
  Abstract         no                                              
  Autowired        no                                              
  Autoconfigured   no                                              
 ---------------- ------------------------------------------------ 
```

If i put a `dump()` inside the constructor of this class `XYZApiProxy`
and launching `symfony console app:admin:update-password` the dump will be printed on the terminal, and that class has nothing to do with the command

I do not know if this helps understanding more the bug I am facing

**Comment by stof** [comment]
The question is where this XYZApiProxy is injected. Maybe it is injected in some dependency of the EntityManager (in a doctrine listener maybe) or of the UserMailer.
Another option is that it is in a dependency graph of a different command not supporting lazy registration (lazy registration is supported either by defining the name through `static $defaultName` or by putting the `alias` attribute on the `console.command` tag in the service definition).

**Comment by noniagriconomie** [comment]
@stof thank you for taking time to on this

I will check but i think all our commands are made lazy with the `static $defaultName` already.

The XYZApiProxy is a service to a payment provider. It is a service that have n dependencies, some are marked lazy true, others not.
The fact is that we do have another payment provider, also with dependencies but none lazy, and if I add a `dump` in the constructor the same `dump` occurs.

---

Are we ok to say that it should not be the case via the container
and that there is something wrong occuring (that may be specific on our project) ?

thank again

**Comment by noniagriconomie** [comment]
@stof @dmaicher

After further investigations, it appears that we have recently introduced an internal bundle that have its dependcyinjection part, and somes classes were declared/instanciated wrongly via config inside, and thus causing the fail.

I think this issue can be considered as resolved 

---

Just a question, for searching to understand the container, I wanted to have a visualization of it, and I found the `Symfony\Component\DependencyInjection\Dumper\GraphvizDumper` via https://www.orbitale.io/2018/12/04/the-symfony-container-graph.html

Is there a reason why there is not a command like https://symfony.com/doc/current/workflow/dumping-workflows.html for example?
(I did not search PRs neither issues on the main symfony repo to check if someone already proposed it etc)

Thank you!
